src/traces/scattergl/index.js

```diff
@@ -823,49 +823,91 @@ function calcHover(pointData, x, y, trace) {
+    var i;
 
-    if(!scene) return selection;
+    if(!scene) return [];
 
     var hasText = subTypes.hasText(trace);
     var hasMarkers = subTypes.hasMarkers(trace);
     var hasOnlyLines = !hasMarkers && !hasText;
-    if(trace.visible !== true || hasOnlyLines) return selection;
+    if(trace.visible !== true || hasOnlyLines) return [];
 
-    // degenerate polygon does not enable selection
-    // filter out points by visible scatter ones
-    var els = null;
-    var unels = null;
-    // FIXME: clearing selection does not work here
-    var i;
     if(polygon !== false && !polygon.degenerate) {
-        els = [], unels = [];
         for(i = 0; i < stash.count; i++) {
             if(polygon.contains([stash.xpx[i], stash.ypx[i]])) {
-                els.push(i);
-                selection.push({
-                    pointNumber: i,
-                    x: x[i],
-                    y: y[i]
-                });
-            }
-            else {
-                unels.push(i);
+                pointsIn.push(i);
             }
         }
+    }
+
+    return pointsIn;
+}
+
+function toggleSelected(searchInfo, selected, pointIds) {
+    var clearSelection = selected === false && !Array.isArray(pointIds);
+    var selection = [];
+    var calcData = searchInfo.cd;
+    var stash = calcData[0].t;
+    var scene = stash._scene;
+    var trace = calcData[0].trace;
+    var hasText = subTypes.hasText(trace);
+    var hasMarkers = subTypes.hasMarkers(trace);
+    var oldEls;
+    var oldUnels;
+    var newEls;
+    var newUnels;
+    var i;
+
+    if(!Array.isArray(pointIds)) {
+        pointIds = arrayRange(stash.count);
+    }
+
+    // Mutate state
+    coerceSelectBatches(scene, stash, hasMarkers);
+    oldEls = scene.selectBatch[stash.index];
+    oldUnels = scene.unselectBatch[stash.index];
+
+    if(selected) {
+        newEls = oldEls.concat(Lib.difference(pointIds, oldEls));
+        newUnels = Lib.difference(oldUnels, pointIds);
     } else {
-        unels = arrayRange(stash.count);
+        newEls = Lib.difference(oldEls, pointIds);
+        newUnels = oldUnels.concat(Lib.difference(pointIds, oldUnels));
+    }
+
+    scene.selectBatch[stash.index] = clearSelection ? null : newEls;
+    scene.unselectBatch[stash.index] = clearSelection ? arrayRange(stash.count) : newUnels;
+
+    // Compute and return selection array from internal state
+    for(i = 0; i < newEls.length; i++) {
+        selection.push({
+            pointNumber: newEls[i],
+            x: stash.x[newEls[i]],
+            y: stash.y[newEls[i]]
+        });
+    }
+
+    if(hasText) {
+        styleTextSelection(calcData);
     }
 
-    // make sure selectBatch is created
+    // TODO remove eventually
+    // console.log('els  : ' + JSON.stringify(scene.selectBatch[stash.index]));
+    // console.log('unels: ' + JSON.stringify(scene.unselectBatch[stash.index]));
+    // console.log('selection: ' + JSON.stringify(selection));
+    return selection;
```

Comment:
I have some bad news. The new selection algo perform very poorly above 1e5 points (which in practice only affects `scattergl`). From

```js
var N = 1e5
var x = []
var y = []
var ms = []

for(var i = 0; i < N; i++) {
  x.push(Math.random())
  y.push(Math.random())
  ms.push(i % 2 ? 20 : 10)
}

console.time(N)
Plotly.newPlot(gd, [{
  type: 'scattergl',
  mode: 'markers',
  x: x,
  y: y,
  marker: {size: ms}
}], {
  dragmode: 'select',
})
console.timeEnd(N)
```

selecting is painfully slow.

----

I'll try to pin down what's causing this slow down later today.

Reply:
Okay, thanks! Let me know if I can be of any assistance, e.g. if some of my code isn't clear...